RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h

```diff
@@ -0,0 +1,184 @@
+#ifndef RecoEgamma_EgammaTools_EnergyScaleCorrection_h
+#define RecoEgamma_EgammaTools_EnergyScaleCorrection_h
+
+//author: Alan Smithee
+//description: 
+//  A port of Shervin Nourbakhsh's EnergyScaleCorrection_class in EgammaAnalysis/ElectronTools
+//  this reads the scale & smearing corrections in from a text file for given categories
+//  it then allows these values to be accessed
+
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <cmath>
+#include <string>
+#include <bitset> 
+
+
+class EnergyScaleCorrection
+{	
+public:  
+  enum FileFormat{
+    UNKNOWN=0,
+    GLOBE,
+    ECALELF_TOY,
+    ECALELF
+  };
+
+  enum ParamSmear{
+    kNone = 0,
+    kRho,
+    kPhi,
+    kNParamSmear
+  };
+  
+  enum ScaleNuisances{
+    kErrStatBitNr = 0,
+    kErrSystBitNr = 1,
+    kErrGainBitNr = 2,
+    kErrNrBits=3,
+    kErrNone = 0,
+    kErrStat = 1,
+    kErrSyst = 2,
+    kErrGain = 4,
+    kErrStatSyst = 3,
+    kErrStatGain = 5,
+    kErrSystGain = 6,
+    kErrStatSystGain = 7
+  };
+  
+  
+
+  
+  class ScaleCorrection
+  {
+  public:  
+    ScaleCorrection():
+      scale_(1.),scaleErrStat_(0.),scaleErrSyst_(0.),scaleErrGain_(0.){}
+    ScaleCorrection(float iScale,float iScaleErrStat,float iScaleErrSyst,float iScaleErrGain):
+      scale_(iScale),scaleErrStat_(iScaleErrStat),scaleErrSyst_(iScaleErrSyst),scaleErrGain_(iScaleErrGain){}
+    
+    float scale()const{return scale_;}
+    float scaleErr(const std::bitset<kErrNrBits>& uncBitMask)const;
+    float scaleErrStat()const{return scaleErrStat_;}
+    float scaleErrSyst()const{return scaleErrSyst_;}
+    float scaleErrGain()const{return scaleErrGain_;}
+
+    friend std::ostream& operator<<(std::ostream& os, const ScaleCorrection& a){return a.print(os);}
+    std::ostream& print(std::ostream& os)const;
+  private:    
+    float scale_, scaleErrStat_, scaleErrSyst_, scaleErrGain_;
+  };
+  
+  struct SmearCorrection
+  {
+  public:  
+    SmearCorrection():
+      rho_(0.),rhoErr_(0.),phi_(0.),phiErr_(0.),
+      eMean_(0.),eMeanErr_(0.){}
+    SmearCorrection(float iRho,float iRhoErr,float iPhi,float iPhiErr,float iEMean,float iEMeanErr):
+      rho_(iRho),rhoErr_(iRhoErr),phi_(iPhi),phiErr_(iPhiErr),
+      eMean_(iEMean),eMeanErr_(iEMeanErr){}
+    
+    friend std::ostream& operator<<(std::ostream& os, const SmearCorrection& a){return a.print(os);}
+    std::ostream& print(std::ostream& os)const;
+   
+   
+    float sigma(const float et,const float nrSigmaRho=0.,const float nrSigmaPhi=0.)const{
+      const double rhoVal = rho_ + rhoErr_ * nrSigmaRho;
+      const double phiVal = phi_ + phiErr_ * nrSigmaPhi;
+      const double constTerm =  rhoVal * std::sin(phiVal);
+      const double alpha =  rhoVal *  eMean_ * std::cos(phiVal);
+      return std::sqrt(constTerm * constTerm + alpha * alpha / et);
```

Comment:
dimensions do not match. Please check. If the formula is "correct" please explain the meaning of it, it has "1 + (x GeV)"

Reply:
the math can be reduced and can likely be all float

Reply:
thanks for the check here. However the formula appears to be correct

sigma^2 = constTerm^2 +  (alpha/sqrt(Et) )^2

Assuming the alpha is related to the stochastic term. 
I need to double check with @shervin86 why its Et not E but its clear its what he intends from 

https://twiki.cern.ch/twiki/bin/view/CMS/EGMSmearer#How_to_compute_systematics_uncer

Math is all float now  but I do kinda like it being explicit like this to avoid readability issues.

Reply:
well, it's sigma^2 = rho^2*(sin2 + cos2*eMean^2/et), which looks particularly odd, as if eMean magically has a dimension of GeV^0.5, unless it's the result of a cos, which is dimensionful.